Relax BALC[32] to 16-bit variant using trampolines

djtodoro opened the pull request:
Enabled by default, to disable please use `--no-relax-balc-trampolines`.

The patch on the LLVM side: https://github.com/MediaTek-Labs/llvm-project/pull/25

farazs-github committed:
[Bug #39] Expand ADDIU[GP.B] to ADDIU[GP.W]

gold/
	* nanomips-insn.def: Add new transformation.
	* nanomips.cc (Nanomips_expand_insn::type): Expand ADDIU[GP.B]
	to ADDIU[GP.W] if the offset is word-aligned and fits 21 bits.
	* testsuite/Makefile.am: Add new tests.
	* testsuite/Makefile.in: Re-generate.
	* testsuite/nanomips_gprel_out_of_range_small.s: New source.
	* testsuite/nanomips_gprel_out_of_range_small.t: New script.
	* testsuite/nanomips_gprel_out_of_range.sh: Check new test
	output.

motokultivator committed:
Relax BALC[32] to 16-bit variant using trampolines

Enabled by default, to disable use
  --no-relax-balc-trampolines

farazs-github commented:
Just for my reference, what kind of integration testing has this patch seen?

djtodoro commented:
> Just for my reference, what kind of integration testing has this patch seen?

- `check-all` passes
- built `llvm-test-suite` (no regression nor improvements)

farazs-github commented:
No changes in llvm-test-suite sounds suspect. This was with the accompanying llvm patch and -Os?

djtodoro commented:
Well, I agree.  I will mark this as DRAFT until I get back with the feedback.

farazs-github commented:
Well, if you are revising the patch, 2 points:
1. We discussed that the placeholder relocation would be added to those calls that we do *not* want to be optimized. Since most of the code is expected to be compiled with -Os and have the optimization enabled, it would be more efficient to mark the ones we want to skip.
2. Lets decide on a relocation name instead of NONE ... something to indicate that it is a non-stub/non-tramp call, something like R_NANOMIPS_NOTRAMP or R_NANOMIPS_DIRECT_CALL. The relocation number will be 77. I don't mind reviewing with R_NANOMIPS_NONE, but I am pretty sure we won't merge the patch without this change.

djolertrk committed:
Add R_NANOMIPS_NOTRAMP

farazs-github commented on gold/nanomips.cc:
```diff
@@ -5863,6 +6064,102 @@ Target_nanomips<size, big_endian>::do_relax(
       // Change the state to EXPAND if we are done with relaxations.
       if (!again && (this->state_ == RELAX) && parameters->options().expand())
         this->state_ = EXPAND;
+      // Change the state to TRAMPOLINE.
```
How do you tackle interactions between the expansion phase and the trampoline phase - so for example what happens when a PC-relative reference that was in-range during expansion goes out-of-range due to the introduction of stubs? This can happen for an LAPC[32] or BALC[16] referring to a different section as an example.

I'd expect these transforms to either be performed together in one pass, or in a loop with one feeding off the other, until things stabilize.

djtodoro replied:
Yes, you are right. We are doing `expand` after we are done with producing the `trampolines`.

farazs-github commented on gold/nanomips.cc:
```diff
@@ -5863,6 +6064,102 @@ Target_nanomips<size, big_endian>::do_relax(
+                  Balc_trampoline_target &t = targets[titer->second];
+                  Address address = balc_trampolines_[i].address;
+                  Address first = balc_trampolines_[t.first].address;
+                  if (t.trampoline == -1ull && (address - 1024 >= first))
```
Style: Add parenthesis block for the nested if-else

djolertrk committed:
Address comments

djolertrk committed:
Fix finding R_NANOMIPS_PC10_S1 target

djolertrk committed:
Suppress warning for unused vars

farazs-github commented on gold/nanomips-insn.def:
```diff
@@ -235,6 +241,8 @@ NTT(PCREL16,          RELS(R(PC25_S1)),                               INSNS(BALC
 NTT(PCREL16_LONG,     RELS(R(PC25_S1)),                               INSNS(ALUIPC32(PC_HI20, SREG), ORI32(LO12, SREG, SREG), JALRC16(SREG)))
 NTT(PCREL32_LONG,     RELS(R(PC25_S1)),                               INSNS(ALUIPC32(PC_HI20, SREG), ORI32(LO12, SREG, SREG), JALRC32))
+NTT(BALC_TRAMP,       RELS(R(PC25_S1)),                               INSNS(BALC16_FIXED, BC16_FIXED, BC32))
+NTT(BALC_CALL,        RELS(R(PC25_S1)),                               INSNS(BALC16))
```
I don't see where the BALC_CALL transformation type is actually affected by changing the  the call-site to go to the trampoline instead of the final function. You can't just be changing the BALC32 to BALC16, because the next expansion phase will just expand it right back to BALC32 - this is indeed what seems to be happening looking at the `--debug=target' log output. You need to change the relocation symbol as well.

motokultivator replied:
This has been fixed in the last few commits, and llvm-test suite now passes.
I'll also add some test cases soon to the project.

farazs-github reviewed (changes requested):
I suggest you evaluate this against some simple test cases, before resubmitting. I found it to be broken in the most obvious way, by inspecting a few random tests from llvm-test-suites (gcc-loops, sradKernel). Calls that get contracted in the Trampoline phase get expanded back in a later Expansion phase, except the ones where you are inserting the trampoline with an extra BC.

Here is a simple and obvious test:
	.text
	.linkrelax
	.module pcrel
	.globl test
	.type test,@function
	.ent test
test:
	jrc $ra
	.rept 512
	nop32
	.endm
	.end test

	.globl main
	.type main,@function
	.ent main
main:
	balc test
	nop
	balc test
	nop
	balc test
	nop
	balc test
	nop
	balc test
	nop
	.end main

motokultivator committed:
Enable additional expansions after the trampoline

motokultivator committed:
Redefine relaxation state loop

motokultivator committed:
Consider r_addend for the current relocation

motokultivator committed:
Add testcase for --relax-balc-trampolines

farazs-github commented on gold/nanomips.cc:
```diff
@@ -193,6 +193,32 @@ is_forced_insn_length(typename elfcpp::Elf_types<size>::Elf_Addr offset,
+  const int reloc_size = elfcpp::Elf_sizes<size>::rela_size;
+
+  preloc += reloc_size;
+  for (size_t i = relnum + 1; i < reloc_count; ++i, preloc += reloc_size)
```
Note for future improvement: reloc list is probably pre-sorted and amenable to a faster search.

farazs-github commented on gold/nanomips.cc:
```diff
@@ -1861,19 +1924,72 @@ class Target_nanomips : public Sized_target<size, big_endian>
+  {
+    static size_t pos = 0;
+    size_t sz = balc_trampolines_.size();
+    for (size_t i = 0; i < sz; ++i)
```
For future improvement: this requests would appear in a roughly ordered manner, so a sliding window would be ideal to limit this search.

motokultivator replied:
This has been addressed.

farazs-github commented on gold/nanomips.cc:
```diff
@@ -7297,6 +7653,10 @@ Target_nanomips<size, big_endian>::scan_reloc_section_for_transform(
                                                   i, prelocs))
         continue;
 
+      unsigned int notramp_reloc =
```
Incorrect type! why not bool?

farazs-github replied:
Incorrect type? Why not bool?

farazs-github commented on gold/nanomips.cc:
```diff
@@ -7297,6 +7653,10 @@ Target_nanomips<size, big_endian>::scan_reloc_section_for_transform(
 
+      unsigned int notramp_reloc =
+        has_notramp_reloc<size, big_endian>(r_offset, reloc_count, i, prelocs);
+      bool has_balc_stub2 = notramp_reloc == false;
```
1. Parenthesis around the comparison expression.

2. From my understanding, please confirm if this is right:
Input: balc[32] with NOTRAMP -> no_tramp_reloc == true -> has_balc_stub2 == false
Input: balc[32] without NOTRAMP -> no_tramp_reloc == false -> has_balc_stub2 == true
Input: any other instruction without NOTRAMP -> no_tramp-reloc == false -> **has_balc_stub2 == true**??
The question is, why is has_balc_stub2 true at all for instructions that are not 32-bit calls? At the least, the variable name is misnamed!

farazs-github commented on gold/nanomips.cc:
```diff
@@ -5598,6 +5753,79 @@ Nanomips_expand_insn_finalize<size, big_endian>::type(
+{
+  unsigned int r_type = elfcpp::elf_r_type<size>(reloc.get_r_info());
+
+  if ((r_type != elfcpp::R_NANOMIPS_PC25_S1 || !has_balc_stub2)
```
Unnecessary nesting. Would be better to put all parts of the || expression at the same level.

Also note the comment about the name `has_balc_stub2' below. I'd expect has_balc_stub2 to imply that it is a 32-bit call instruction and already has the appropriate relocation and any checking to that effect would simply be in an assert( ).

motokultivator committed:
Add design doc

AndrijaSyrmia committed:
Just added .vscode in gitignore

AndrijaSyrmia committed:
Implementation of balc stubs fix + one test for testing balc stubs

Fixed issues in align

Changed picking balc tramp candidates a bit

Reimplemented processing of R_NANOMIPS_NOTRAMP
